Fix error message from new symbolic link missing target

yecril71pl opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->
Fixes #13073 (well, at least descreases its morbidity).

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->
When the target value was not given and `New-Item` throws an exception, use the custom message format intended to be used.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

yecril71pl committed:
when a symbolic link target is not specified

Display a meaningful error message when a symbolic link target is not specified!

yecril71pl committed:
NewArgumentNullException accepts format at 1

Fix: NewArgumentNullException accepts format parameter at position 1.

yecril71pl committed:
Add block wrapper for Should -Throw

iSazonov commented on src/System.Management.Automation/engine/SessionStateContainer.cs:
```diff
@@ -3500,14 +3500,14 @@ internal Collection<PSObject> NewItem(string[] paths, string name, string type,
                         if (content == null)
                         {
-                            throw PSTraceSource.NewArgumentNullException(SessionStateStrings.NewItemValueNotSpecified, path);
+                            throw PSTraceSource.NewArgumentNullException(nameof(content), SessionStateStrings.NewItemValueNotSpecified, path);
```
Please update the resource string. It is used only in the context so we can get better message like:
```
New-Item: Item '.' cannot be created because Target was not specified.
```

iSazonov commented on src/System.Management.Automation/engine/SessionStateContainer.cs:
```diff
@@ -3500,14 +3500,14 @@ internal Collection<PSObject> NewItem(string[] paths, string name, string type,
                         if (string.IsNullOrEmpty(targetPath))
                         {
-                            throw PSTraceSource.NewArgumentNullException(SessionStateStrings.PathNotFound, targetPath);
+                            throw PSTraceSource.NewArgumentNullException(nameof(targetPath), SessionStateStrings.PathNotFound, targetPath);
```
I think it was and is not correct.
```suggestion
                            throw PSTraceSource.NewArgumentNullException(nameof(targetPath), SessionStateStrings.NewItemValueNotSpecified, path);
```

yecril71pl replied:
I think it is dead code.

iSazonov replied:
Really the New-Item parameter is `Value` and type of the parameter is `object`. So user can set an object here. I agree that it is an edge case but for consistency we should fix and test the code path. (You could thing about that the Value could get more complicated expression in future, maybe scriptblock?)

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -185,6 +185,11 @@ Describe "New-Item" -Tags "CI" {
+
+    It "Should display an error message when a symbolic link target is not specified" {
+        { New-Item -Name $testfile -Path $tmpDirectory -ItemType SymbolicLink } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
+        $ERROR | Select-Object -First 1 | Foreach-Object Exception | Foreach-Object Message | Should -Contain 'value'
```
Please remove this. We never check "Message" because of localization.
We could check `ParamName` instead.

Also we could add a test for second condition too (for `content.ToString()`)

yecril71pl committed:
Simplify empty value handler

The error message should be the same whether the value is null or empty.

yecril71pl committed:
Check missing content parameter identified

I was told we cannot test the exception message because of localisation and that we should test the missing parameter instead.  I disagree because CI tests run under C locale anyway, byt here you are.

yecril71pl committed:
improve code style

yecril71pl committed:
NewItemValueNotSpecified be …LinkTarget…

This message is currently used when we fail to create a link, so we decided to be more specific.  I also changed the name to reflect this assumption.

yecril71pl committed:
Rename NewItemValueNotSpecified

The new name is NewLinkTargetNotSpecified.

yecril71pl committed:
Fix stupid typo

iSazonov commented on src/System.Management.Automation/engine/SessionStateContainer.cs:
```diff
@@ -3498,16 +3498,11 @@ internal Collection<PSObject> NewItem(string[] paths, string name, string type,
+                        string targetPath;
 
-                        if (string.IsNullOrEmpty(targetPath))
+                        if (content == null || string.IsNullOrEmpty(targetPath = content.ToString()))
```
Please use new operator:
```suggestion
                        if (content is null || string.IsNullOrEmpty(targetPath = content.ToString()))
```

yecril71pl committed:
Style: is null

Co-authored-by: Ilya <darpa@yandex.ru>

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -185,6 +185,13 @@ Describe "New-Item" -Tags "CI" {
     }
+
+    It "Should display an error message when a symbolic link target is not specified" {
+        { New-Item -Name:$testfile -Path:$tmpDirectory -ItemType:SymbolicLink } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
+        $Error | Select-Object -First:1 | Foreach-Object Exception | Foreach-Object ParamName | Should -Be 'content'
+        { New-Item -Name:$testfile -Path:$tmpDirectory -ItemType:SymbolicLink -Value:{} } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
+        $Error | Select-Object -First:1 | Foreach-Object Exception | Foreach-Object ParamName | Should -Be 'content'
```
Please follow common test style.
We can safely use `$Error[0]`.
We prefer to use `-BeExactly` for strings.
```suggestion
        { New-Item -Name $testfile -Path $tmpDirectory -ItemType SymbolicLink } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
        $Error[0].Exception.ParamName | Should -BeExactly 'Content'
        { New-Item -Name $testfile -Path $tmpDirectory -ItemType:SymbolicLink -Value {} } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
         $Error[0].Exception.ParamName | Should -BeExactly 'Content'
```

yecril71pl committed:
follow common test style

Co-authored-by: Ilya <darpa@yandex.ru>

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -185,6 +185,13 @@ Describe "New-Item" -Tags "CI" {
+
+    It "Should display an error message when a symbolic link target is not specified" {
+        { New-Item -Name $testfile -Path $tmpDirectory -ItemType SymbolicLink } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
+        $Error[0].Exception.ParamName | Should -BeExactly 'Content'
+        { New-Item -Name $testfile -Path $tmpDirectory -ItemType:SymbolicLink -Value {} } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
+         $Error[0].Exception.ParamName | Should -BeExactly 'Content'
```
```suggestion
        $Error[0].Exception.ParamName | Should -BeExactly 'content'
        { New-Item -Name $testfile -Path $tmpDirectory -ItemType:SymbolicLink -Value {} } | Should -Throw -ErrorId 'ArgumentNull,Microsoft.PowerShell.Commands.NewItemCommand'
         $Error[0].Exception.ParamName | Should -BeExactly 'content'
```

iSazonov committed:
Update test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1

iSazonov commented:
@yecril71pl Thanks for discovering and fixing the issue!

yecril71pl commented:
> @yecril71pl Thanks for discovering and fixing the issue!

My pleasure.  I would like to note that:
1. I am still unhappy that `-TARGET` is not formally mandatory in this case.
2. I contend that `-PARAMETER:VALUE` is a better syntax in most cases because it is explicit and unambiguous.

iSazonov commented:
> I am still unhappy that -TARGET is not formally mandatory in this case.

The cmdlet is provider based. I think it is impossible to make it mandatory for all scenarios. So the fix is good.

> I contend that -PARAMETER:VALUE is a better syntax in most cases because it is explicit and unambiguous

We follow common pattern for tests. It would be huge work to change a style for all tests.
